// summer-mvc/tools/db/mysqlHandler.js
module.exports.getPool = getPool;
module.exports.getConnection = getConnection;
module.exports.executeQuery = executeQuery;
module.exports.releasePool = releasePool;
module.exports.releaseConnection = releaseConnection;
module.exports.getQueries = getQueries;





const mysql = require( "mysql" );
const queries = require("fs").existsSync( __mysqlQueries ) === true ? require( __mysqlQueries ) : null;





function getPool(){
	return new Promise( function(resolve, reject){
		try{
			resolve( mysql.createPool( require(__mysqlHandlerInfo) ) );
		} catch( err ){
			console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
			reject( err );
		}
	} );
}

function getConnection(){
	try {
		var _pool = undefined;

		if( arguments ){
			_pool = arguments[0];
		}

		return new Promise( function(resolve, reject){

			if( _pool ){
				_pool.getConnection( function(err, _connection){
					if( err ) reject( err );
					resolve( _connection );
				} );

			} else{
				getPool()
				.then( function(_pool){
					_pool.getConnection( function(err, _connection){
						if( err ) reject( err );
						resolve( _connection );
					} );
				} );
			}
		} );
	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
		throw err;
	}
}






function executeQuery( queryId ){
	try {
		var params = undefined;
		var connection = undefined;

		if( arguments[1] && typeof arguments[1] == "object" ){
			if( arguments[1] instanceof Array ){
				params = arguments[1];

				if( arguments[2] ){
					connection = arguments[2];
				}
			} else if( !(arguments[1] instanceof Array) ){
				connection = arguments[1];
			}
		}

		return new Promise( function(resolve, reject){
			if( connection ){
				executeTransaction( queryId, params, connection )
				.then( function( queryResults ){
					resolve( queryResults );
				} )
				.catch( function(err){
					console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
					reject( err );
				} );
			} else{
				getConnection()
				.then( executeTransaction.bind(null, queryId, params) )
				.then( function(queryResults){
					releaseConnection( queryResults.connection );
					releasePool( queryResults.connection.config.pool );

					resolve( queryResults );
				} )
				.catch( function(err){
					console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
					reject( err );
				} );
			}
		} );
	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
		throw err;
	}
}

function executeTransaction( queryId, params, connection ){

	return new Promise( function(resolve, reject){
		try{
			connection.beginTransaction( function(err){
				if( err ){
					console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
					reject( err );
				}

				connection.query( getQueryString( queryId, params ), params, function(err, results, fields){

					if( err ){
						connection.rollback( function(){
							console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
							reject( err );
						} );
					}

					connection.commit( function(err){
						if( err ){
							connection.rollback( function(){
								console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
								reject( err );
							} )
						}

						// for( var i=0; i<results.length; i++ ){

						// 	for( key in results[i] ){
						// 		if( typeof ( (results[i] )[key] ) === "string" ){
						// 			( results[i] )[key] = decodeURIComponent( ( results[i] )[key] );
						// 		}
						// 	}
						// }

						resolve( { connection:connection, results:results, fields:fields } );
					});

				} );
			} );
		} catch( err ){
			console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
			reject( err );
		}
	} );
}






function releasePool( pool ){
	try {
		pool.end();
	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
		throw err;
	}
}

function releaseConnection( _connection ){
	try {
		_connection.destroy();
	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
		throw err;
	}
}

function getQueryString( queryId, params ){
	try {
		if( mysqlQueriesXML ){
			let queriesArr = mysqlQueriesXML.queries.query;

			for( var i=0; i<queriesArr.length; i++ ){
				if( queriesArr[i].$.id == queryId ){
					return setQueryParams( queriesArr[i]._, params );
				}
			}
		}
		
		if( queriesXML ){
			let queriesArr = queriesXML.queries.query;

			for( var i=0; i<queriesArr.length; i++ ){
				if( queriesArr[i].$.id == queryId ){
					return setQueryParams( queriesArr[i]._, params );
				}
			}
		}

		for( var i=0; i<queries.length; i++ ){
			if( queries[i].id == queryId ){
				return queries[i].queryString;
			}
		}

		return "SELECT NOW()";
	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
		throw err;
	}
}

function setQueryParams( queryString, params ){
	try {
		if( params ){
			if( queryString.indexOf("?") > -1 ){

				for( var i=0; i<params.length; i++ ){

					if( typeof params[i] === "string" && params[i] != null && params[i] != undefined ){
						params[i] = params[i].replace( /\'/gi, "&#x27;" );	
						params[i] = params[i].replace( /\?/gi, "&#63;" );	
						// params[i] = encodeURIComponent( params[i] );
					}

					if( !isNaN(parseFloat(params[i])) && isFinite(params[i]) ){
						queryString = queryString.replace( "?", "" + params[i] + "" );
					} else {
						queryString = queryString.replace( "?", "'" + params[i] + "'" );
					}

				}
			} else{
				for( var i=0; i<params.length; i++ ){
					if( params[i] instanceof Object ){

						var key = Object.keys( params[i] )[0];
						var value = ( params[i] )[key];

						if( typeof value === "string" && value != null && value != undefined ){
							value = value.replace( /\'/gi, "&#x27;" );
							value = value.replace( /\?/gi, "&#63;" );
							// value = encodeURIComponent( value );
						}
						

						if( !isNaN(parseFloat(value)) && isFinite(value) ){
							queryString = queryString.replace( new RegExp("#" + key + "#", "gi"), value );
						} else {
							queryString = queryString.replace( new RegExp("#" + key + "#", "gi"), "'" + value + "'" );
						}
					}
				}
			}
		}

		return setLikeQueryParam( queryString );
	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mssqlHandler.js]", err );
		throw err;
	}
}

function setLikeQueryParam( queryString ){
	try {
		var queryArray = queryString.split( /where/i );
		var prePhase = queryArray[0];
		var splitPhase = "";

		if( queryArray.length > 1 ){
			for( var j=1; j<queryArray.length; j++ ){
				var wherePhase = ( ( queryArray[j] ).trim() ).replace( /\s\s+/g, ' ' );

				if( wherePhase.length > 0 ){
					var wherePhaseArray = wherePhase.split(" ");

					for( var i=1; i<wherePhaseArray.length; i++ ){

						if( wherePhaseArray[i-1].toUpperCase() === "LIKE" ){
							var param = wherePhaseArray[i];
							param = param.replace( /'/gi, "" );
							param = "'%" + param + "%'";
							wherePhaseArray[i] = param;
						}
					}

					wherePhase = "";

					for( var i=0; i<wherePhaseArray.length; i++ ){
						wherePhase += " " + wherePhaseArray[i];
					}
				}

				splitPhase += " WHERE" + wherePhase;
			}

			queryString = prePhase + splitPhase;
		}

		return queryString;

	} catch( err ){
		console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mssqlHandler.js]", err );
		throw err;
	}
}

function translateParam( param ){
	var chars = ["©","Û","®","ž","Ü","Ÿ","Ý","$","Þ","%","¡","ß","¢","à","£","á","À","¤","â","Á","¥","ã","Â","¦","ä","Ã","§","å","Ä","¨","æ","Å","©","ç","Æ","ª","è","Ç","«","é","È","¬","ê","É","­","ë","Ê","®","ì","Ë","¯","í","Ì","°","î","Í","±","ï","Î","²","ð","Ï","³","ñ","Ð","´","ò","Ñ","µ","ó","Õ","¶","ô","Ö","·","õ","Ø","¸","ö","Ù","¹","÷","Ú","º","ø","Û","»","ù","Ü","@","¼","ú","Ý","½","û","Þ","€","¾","ü","ß","¿","ý","à","‚","À","þ","á","ƒ","Á","ÿ","å","„","Â","æ","…","Ã","ç","†","Ä","è","‡","Å","é","ˆ","Æ","ê","‰","Ç","ë","Š","È","ì","‹","É","í","Œ","Ê","î","Ë","ï","Ž","Ì","ð","Í","ñ","Î","ò","‘","Ï","ó","’","Ð","ô","“","Ñ","õ","”","Ò","ö","•","Ó","ø","–","Ô","ù","—","Õ","ú","˜","Ö","û","™","×","ý","š","Ø","þ","›","Ù","ÿ","œ","Ú"]; 
	var codes = ["&copy;","&#219;","&reg;","&#158;","&#220;","&#159;","&#221;","&#36;","&#222;","&#37;","&#161;","&#223;","&#162;","&#224;","&#163;","&#225;","&Agrave;","&#164;","&#226;","&Aacute;","&#165;","&#227;","&Acirc;","&#166;","&#228;","&Atilde;","&#167;","&#229;","&Auml;","&#168;","&#230;","&Aring;","&#169;","&#231;","&AElig;","&#170;","&#232;","&Ccedil;","&#171;","&#233;","&Egrave;","&#172;","&#234;","&Eacute;","&#173;","&#235;","&Ecirc;","&#174;","&#236;","&Euml;","&#175;","&#237;","&Igrave;","&#176;","&#238;","&Iacute;","&#177;","&#239;","&Icirc;","&#178;","&#240;","&Iuml;","&#179;","&#241;","&ETH;","&#180;","&#242;","&Ntilde;","&#181;","&#243;","&Otilde;","&#182;","&#244;","&Ouml;","&#183;","&#245;","&Oslash;","&#184;","&#246;","&Ugrave;","&#185;","&#247;","&Uacute;","&#186;","&#248;","&Ucirc;","&#187;","&#249;","&Uuml;","&#64;","&#188;","&#250;","&Yacute;","&#189;","&#251;","&THORN;","&#128;","&#190;","&#252","&szlig;","&#191;","&#253;","&agrave;","&#130;","&#192;","&#254;","&aacute;","&#131;","&#193;","&#255;","&aring;","&#132;","&#194;","&aelig;","&#133;","&#195;","&ccedil;","&#134;","&#196;","&egrave;","&#135;","&#197;","&eacute;","&#136;","&#198;","&ecirc;","&#137;","&#199;","&euml;","&#138;","&#200;","&igrave;","&#139;","&#201;","&iacute;","&#140;","&#202;","&icirc;","&#203;","&iuml;","&#142;","&#204;","&eth;","&#205;","&ntilde;","&#206;","&ograve;","&#145;","&#207;","&oacute;","&#146;","&#208;","&ocirc;","&#147;","&#209;","&otilde;","&#148;","&#210;","&ouml;","&#149;","&#211;","&oslash;","&#150;","&#212;","&ugrave;","&#151;","&#213;","&uacute;","&#152;","&#214;","&ucirc;","&#153;","&#215;","&yacute;","&#154;","&#216;","&thorn;","&#155;","&#217;","&yuml;","&#156;","&#218;"];

	for(x=0; x<chars.length; x++){
		for (i=0; i<arguments.length; i++){
		    arguments[i].value = arguments[i].value.replace(chars[x], codes[x]);
		}
	}
}




function getQueries(){
	return new Promise( function(resolve, reject){
		var mysqlHandlerInfo = require(__mysqlHandlerInfo);


		if( "queries" in mysqlHandlerInfo && mysqlHandlerInfo.queries != null ){
			var queryFiles = mysqlHandlerInfo.queries;
			var queryStrings = [];

			var promises = [];

			for( var i=0; i<queryFiles.length; i++){
				promises.push( parsingQueries(queryFiles[i]) );
			}

			Promise.all( promises )
			.then( function(){
				var argv = arguments[0];

				for( var j=0; j<argv.length; j++ ){
					queryStrings = queryStrings.concat( argv[j] );
				}

				var queries = { "queries" : {"query" : queryStrings } };
				
				resolve( queries );
			} )
			.catch( function(_err){
				console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", _err );
				reject( _err );
			} );
		} else {
			parsingQueries( __mysqlQueriesXML)
			.then( function( queryStrings ){
				resolve( { "queries" : {"query" : queryStrings } } );
			} )
			.catch( function(err){
				console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
				reject( err );
			} );
		}	
	} );
}

function parsingQueries( queryFile ){
	return new Promise( function(resolve, reject){
		var fs = require('fs'), xml2js = require('xml2js');
		var parser = new xml2js.Parser();

		if( queryFile.indexOf( process.cwd() ) < 0 ){
			queryFile = require("path").join( process.cwd(), queryFile );
		}

		fs.readFile( queryFile, function(err, data){
			if( err ){
				console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
				reject( err );
			} else {
				parser.parseString(data, function(_err, result){
					if( _err ){
						console.log( "\x1b[31m%s\x1b[0m", "[summer-mvc core]", "[mysqlHandler.js]", err );
						reject( _err );						
					}
					resolve( result.queries.query );
				} );
			}
		} );
	} );
}
